app/modules/net_info.c

```diff
@@ -0,0 +1,133 @@
+// ***************************************************************************
+// net_info module for ESP8266 with nodeMCU
+//
+// adopted jan 2017 for commit to nodeMCU by Wolfgang Rosner
+// re-worked by Lukas Voborsky, @voborsky
+// ***************************************************************************
+
+// #define NODE_DEBUG
+ 
+#include "module.h"
+#include "lauxlib.h"
+#include "platform.h"
+#include "mem.h"
+#include "task/task.h"
+
+#include "lwip/ip_addr.h"
+#include "espconn.h"
+#include "lwip/dns.h" 
+#include "lwip/app/ping.h"
+#include "lwip/raw.h"
+
+/* 
+ping_opt needs to be the first element of the structure. It is a workaround to pass the 
+calllback reference to the ping_received function. Pointer the ping_option structure is
+equal to the pointer to net_info_ping_t strcture. Reference to callback function is assumed
+to be just after the ping_option structure.
+*/
+typedef struct {
+  struct ping_option ping_opt; 
+  sint32_t ping_callback_ref; 
+  uint32 count;
+ } net_info_ping_t;
+typedef net_info_ping_t* ping_t;  
+  
+static void ping_received(void *arg, void *data) {
+    NODE_DBG("[ping_received] arg = %p, data = %p\n", arg, data);
+    ping_t nip  = (ping_t)arg; // override ping_option with net_info_ping_t to access the callback reference
+    struct ping_resp *pingresp = (struct ping_resp*)data; 
+
+    NODE_DBG("[ping_received] count = %d;arg: total_count=%d resp_time=%d seqno=%d timeout_count=%d bytes=%d total_bytes=%d total_time=%d ping_err=%d\n", nip->count, pingresp->total_count, pingresp->resp_time, pingresp->seqno, pingresp->timeout_count, pingresp->bytes, pingresp->total_bytes, pingresp->total_time, pingresp->ping_err);
+    
+    char ipaddrstr[16];
+    ip_addr_t source_ip;
+    source_ip.addr = nip->ping_opt.ip;
+    ipaddr_ntoa_r(&source_ip, ipaddrstr, sizeof(ipaddrstr));
+
+    lua_State *L = lua_getstate();
+    lua_rawgeti(L, LUA_REGISTRYINDEX, nip->ping_callback_ref);
+    lua_pushinteger(L, pingresp->bytes);
+    lua_pushstring(L, ipaddrstr);
+    lua_pushinteger(L, pingresp->seqno);
+    lua_pushinteger(L, pingresp->resp_time);
+    
+    (nip->count)--;
+    
+    lua_call(L, 4, 0);
+}
+
+static void ping_sent(void *arg, void *data) {
```

Comment:
I am *astounded* that `ping_sent` is called after all the `ping_received` but so it goes.  In which case, it looks like the reference counting issue is indeed fixed.